### DISCUSSION TITLE
Can I set the key of encrypted files during application running

### DISCUSSION THREAD

**Discussion by IceCY** [gramineproject/gramine#1302] [General]
Hi, everyone,

According to the docs, it seems that we should set the key of encrypted files (for specific key_name) before application running. However, my project may need to get the key from the remote during application running.

I am wondering whether it is possible set the key of encrypted files during application running. If it possible, how can I do this?

**Comment by dimakuv** [comment]
> I am wondering whether it is possible set the key of encrypted files during application running. If it possible, how can I do this?

Yes, it is possible. Gramine exposes pseudo-files under `/dev/attestation/keys/` directory. See the very last paragraph here: https://gramine.readthedocs.io/en/latest/attestation.html#low-level-dev-attestation-interface

Here is a sample C application that writes/reads these pseudo-files: https://github.com/gramineproject/gramine/blob/master/libos/test/regression/keys.c

Here is another example (our Secret Provisioning library):https://github.com/gramineproject/gramine/blob/a01265b22b20abe47f11968a2e35318866e5166f/tools/sgx/ra-tls/secret_prov_attest.c#L416-L446

I've seen people writing the same on Google Go and Python languages, but unfortunately I cannot find any link to such examples. But it's definitely possible.
